TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.PerformCallRelation.cs

```diff
@@ -0,0 +1,204 @@
+            /// <param name="performCaller"></param>
+            /// <param name="performCallees"></param>
+            /// <param name="procedures"></param>
+            internal PerformCallChain(PerformProcedure performCaller, List<Node> performCallees, HashSet<Procedure> procedures)
+            {
+                this.PerformCaller = performCaller;
+                this.PerformCallees = performCallees;
+                this.Procedures = procedures;
+            }
+        }
+
+        /// <summary>
+        /// Representation of Perform Call Relation. 
+        /// It is a dictionary which gives for each Procedure its Perform Call chain.
+        /// </summary>
+        private class PerformCallRelation : Dictionary<Procedure, PerformCallChain>
+        {
+            /// <summary>
+            /// The Relation's domain
+            /// </summary>
+            internal HashSet<Procedure> Domain;
+
+            internal PerformCallRelation()
+            {
+                Domain = new HashSet<Procedure>();
+            }
+            /// <summary>
+            /// Add a PERFORM CALL relation : Caller => Callee (Procedure)
+            /// Example of code:
+            ///        main.
+            ///             perform a
+            ///             goback
+            ///             .
+            ///        a.
+            ///             perform b
+            ///        .
+            /// The relation is: a => b
+            /// </summary>
+            /// <param name="performCaller">PERFORM procedure caller, source instruction: in the example (perform a)</param>
+            /// <param name="caller">Caller procedure: in the example 'a'</param>/// 
+            /// <param name="performCallee">The PERFORM procedure called, target instruction: in the example (perform b)</param>
+            /// <param name="callee">The called procedure: in the example 'b' </param>
+            internal void AddToRelation(PerformProcedure performCaller, Procedure caller, PerformProcedure performCallee, Procedure callee)
+            {
+                if (!TryGetValue(caller, out var performCallChain))
+                {
+                    performCallChain = new PerformCallChain(performCaller);
+                    Add(caller, performCallChain);
+                }
+                performCallChain.Procedures.Add(callee);
+                if (!performCallChain.PerformCallees.Contains(performCallee))
+                {
+                    performCallChain.PerformCallees.Add(performCallee);
+                }
+                Domain.Add(caller);
+                Domain.Add(callee);
+            }
+
+            /// <summary>
+            /// Compute the transitive closure of the PERFORM CALL relation.
+            /// We browse the graph represented using the Call Relation, then we found during the visit cycles.
+            /// In the same time we compute the call chain.
+            /// </summary>
+            internal void TransitiveClosure()
```

Comment:
Yes this will be an  improvement because IBM compiler reports theses warnings in such situation:

    12  IGYCB7309-W   There may be a loop from the "PERFORM" statement at "PERFORM (line 12.1)" to itself. 
    12  IGYCB7310-W   The "PERFORM" statement at "PERFORM (line 12.1)" cannot reach its exit.

But it seems that qualis does not report such situation also.

But our parser report a CFG error like this:
--- Diagnostics ---
Line 12[13,21] <37, Warning, General> - Warning: Statement '     go to b' located at line 10, column 13 prevents this PERFORM statement to reach its exit.

Ok for such situation I agree, Abstract Interpretation is welcome to be very similar with IBM Compiler.